### ISSUE TITLE
Form controls missing ARIA labels

### ISSUE THREAD

**Issue by yiskah** [tootsuite/mastodon#539]
The search box, toot composition box, and the two timeline regex boxes are all missing ARIA labels which communicate to uses on screen readers what these boxes do. 

The simple fix would be to add to the `<input>` HTML elements a `label='search box'` `label='composition'` etc. this is so users on screen-readers can quickly find and navigate to the element by label.

**Comment by ineffyble** [comment]
Is a label attribute on an input tag common/accepted standard for accessibility?

I thought the standard was adding a <label for="compose">Compose</label> element (which could be hidden with CSS).

**Comment by yiskah** [comment]
So this series of issues came from using WAVE and AInspector to run audits. Some of them were clearly non-applicable (we give every link a unique name describing where it goes because users wrote the links) but the form controls got flagged by both audits for this and this was the suggested solution. It also lines up with what I've read in WAI-ARIA and WCAG

That said, web standards move fast but accessibility standards don't. It could be that using labels is outdated and I don't know because all of the resources out there for reading about web accessibility are outdated. what I've found is most screen readers are different from each other and don't comply with what WCAG says they should use to navigate a page. It's a mess and really what's important is getting the Apple screen reader on iOS and Mac OSX to work correctly since that's what most people use. I have an OSX box I built originally for the sake of testing sites with the screen reader so if desired I can go test with it maybe this weekend

**Comment by yiskah** [comment]
*we can't give unique names

**Comment by ineffyble** [comment]
...my HTML got rendered, see above now - I was asking if better to use a label attribute or label tag

**Comment by marrus-sh** [comment]
I believe the [aria-label](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute) attribute is the accepted standard?

**Comment by LJWatson** [comment]
@yiskah
VoiceOver on ioS/oSX is not the most common screen reader in use. Check out the [WebAIM screen reader survey](http://webaim.org/projects/screenreadersurvey6/) for useful info. That said, these missing labels affect all screen reader users, so definitely a good thing to fix.

@ineffyble
The ```<label>``` element is needed in this instance, not the `label` attribute. It is not a good idea to hide it visually using CSS (or any other mechanism) though.

The label is used to inform screen reader users what is expected, but it is also used by speech recognition users to accurately target the form field by name using voice commands. For this reason the label should be visible.

The ```<label>``` element is the recommended way to provide a label. It can be associated with the ```<input>``` using the `for` and `id` attribute pairing, as follows:
```
<label for="this">What's on your mind?</label>
<input type="text" id="this">
```

If that proves difficult, then the ```<label>``` can be wrapped around the ```<input>``` instead, as follows:
```
<label>What's on your mind?
<input type="text">
</label>
```

The only common exception to providing a visible label, is a search field. This is because the search button (assuming one is present) usually provides enough of a visual aid for speech recognition users, enabling them to target the search field by name. 

In this case the recommended approach is to use the `title` attribute. It is possible to use the `aria-label` attribute in this context, but when used on form fields the `title` attribute has far better browser and screen reader support (including legacy tech).
```
<input type="search" title="Search">
```

**Comment by Gargron** [comment]
We've been using an eslint a11y plugin to point out such issues and adding all sorts of aria labels since, I think all majors issues have been resolved.
